core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1070,6 +1071,70 @@ private ScanToCalcUnifyRule() {
     }
   }
 
+  /**
+   * A {@link SubstitutionVisitor.UnifyRule} that matches a {@link MutableCalc}
+   * which has {@link MutableAggregate} as  child.
+   * We try to compensate the {@link MutableCalc},
+   * then match the query to the top of {@link MutableCalc} in target.
+   */
+  private static class NothingToCalcUnifyRule extends AbstractUnifyRule {
+
+    public static final NothingToCalcUnifyRule INSTANCE =
+        new NothingToCalcUnifyRule();
+
+    private NothingToCalcUnifyRule() {
+      super(any(MutableRel.class),
+          any(MutableCalc.class), 0);
```

Comment:
How about we define it as 
```
    super(operand(query(0)),
          operand(MutableCalc.class, target(0)), 1);
```
It means `query` are equivalent to input of `target` and the only difference is that there is a `Calc` for permutation in the `target`